### PR TITLE
Disable CGO for node e2e test binary on `linux/arm`

### LINKED ISSUE #115675: [Failing Test] build-master
### Which jobs are failing?

master-blocking

- build-master

### Which tests are failing?

Overall build is failing

Screenshot from 2023-02-10 10-42-47

### Since when has it been failing?

2023-02-09 23:09:58 +0000 UTC

### Testgrid link

https://testgrid.k8s.io/sig-release-master-blocking#build-master

### Reason for failure (if possible)

`make: *** [Makefile:487: cross] Error 1
!!! [0210 00:35:55] Call tree:
!!! [0210 00:35:55]  1: build/../build/common.sh:484 kube::build::run_build_command_ex(...)
!!! [0210 00:35:55]  2: build/release.sh:36 kube::build::run_build_command(...)
make: *** [Makefile:374: release] Error 1
level=fatal msg="failed to run: running make release: command /usr/bin/make release did not succeed: !!! [0210 00:35:54] Call tree:\n!!! [0210 00:35:54]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)\nmake[1]: *** [Makefile:92: all] Error 1\nmake: *** [Makefile:487: cross] Error 1\n!!! [0210 00:35:55] Call tree:\n!!! [0210 00:35:55]  1: build/../build/common.sh:484 kube::build::run_build_command_ex(...)\n!!! [0210 00:35:55]  2: build/release.sh:36 kube::build::run_build_command(...)\nmake: *** [Makefile:374: release] Error 1\n" `

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig testing

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind failing-test

#### What this PR does / why we need it:
The upstream master build fails right now with:

```
/usr/local/go/pkg/tool/linux_amd64/link: running arm-linux-gnueabihf-gcc failed: exit status 1
/tmp.k8s/go-link-865684912/go.o: in function `k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/service/ec2.(*ModifyCapacityReservationInput).GoString':
go.go:(.text+0x207ca88): relocation truncated to fit: R_ARM_CALL against `runtime.duffcopy'
go.go:(.text+0x207ca94): relocation truncated to fit: R_ARM_CALL against `runtime.duffcopy'
collect2: error: ld returned 1 exit status
```
Adding a CGO flag `-mlong-calls` for 32bit arm Linux builds did not work and the relocation still failed. This means we now temporarily disable CGO for this binary build to let `make cross` pass.

#### Which issue(s) this PR fixes:

Refers to #115605
Fixes https://github.com/kubernetes/kubernetes/issues/115675
#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

**Comment by saschagrunert** [comment]
cc @palnabarun

**Review by palnabarun — APPROVED** [review]
Let's try this 🚀 

/lgtm
/approve

**Comment by saschagrunert** [comment]
@palnabarun before digging into test-infra. Do you know a presubmit we can test to verify this change?

**Comment by palnabarun** [comment]
I am not sure right off the bat if we have a pre-submit which builds/tests `linux/arm` combination.

**Comment by saschagrunert** [comment]
This should run through. PTAL @kubernetes/sig-release-leads @kubernetes/release-managers 

For approval, PTAL hack/OWNERS: @bentheelder @cblecker @dchen1107 @deads2k @spiffxp @thockin

**Comment by palnabarun** [comment]
Disabling CGO should be safe for the e2e_node test binary if we are not calling any C code from in there.

Shall we add this info in the comment for anyone looking in the future?

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -780,6 +780,20 @@ kube::golang::build_binaries_for_platform() {
...
+    # out for the release notes which may fix relocation issues like:
+    #
+    # /tmp.k8s/go-link-865684912/go.o: in function `k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/service/ec2.(*ModifyCapacityReservationInput).GoString':
+    # go.go:(.text+0x207ca88): relocation truncated to fit: R_ARM_CALL against `runtime.duffcopy'
+    # go.go:(.text+0x207ca94): relocation truncated to fit: R_ARM_CALL against `runtime.duffcopy'
+    #
+    # Ref: https://github.com/kubernetes/kubernetes/issues/115605
+    if [[ "${test}" == k8s.io/kubernetes/test/e2e_node/e2e_node.test && "${platform}" == linux/arm ]]; then
+      kube::log::info "    disabling CGO for binary: ${test}"
+      export CGO_ENABLED=0
```
Comment (cblecker) [inline hack/lib/golang.sh:794]: If this is exported and not locally scoped, I think this will bleed

**Review by cblecker — COMMENTED** [review]
/hold

File: hack/lib/golang.sh (hunk 1 above)
Comment (saschagrunert) [reply hack/lib/golang.sh:794]: Makes sense, removing the `export` should be fine in that case.

**Comment by saschagrunert** [comment]
Giving it another run:
/test pull-kubernetes-cross

File: hack/lib/golang.sh [hunk 2]
```diff
@@ -780,6 +780,20 @@ kube::golang::build_binaries_for_platform() {
...
+    # out for the release notes which may fix relocation issues like:
+    #
+    # /tmp.k8s/go-link-865684912/go.o: in function `k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/service/ec2.(*ModifyCapacityReservationInput).GoString':
+    # go.go:(.text+0x207ca88): relocation truncated to fit: R_ARM_CALL against `runtime.duffcopy'
+    # go.go:(.text+0x207ca94): relocation truncated to fit: R_ARM_CALL against `runtime.duffcopy'
+    #
+    # Ref: https://github.com/kubernetes/kubernetes/issues/115605
+    if [[ "${test}" == k8s.io/kubernetes/test/e2e_node/e2e_node.test && "${platform}" == linux/arm ]]; then
+      kube::log::info "    disabling CGO for binary: ${test}"
+      CGO_ENABLED=0
```
Comment (liggitt) [inline hack/lib/golang.sh:794]: does setting this persist beyond this scope (e.g. does it pollute CGO_ENABLED values for other build calls)?

Comment (liggitt) [reply hack/lib/golang.sh:794]: and should this be set only for the invocation of the `go test` call immediately below?

File: hack/lib/golang.sh [hunk 3]
```diff
@@ -780,6 +780,20 @@ kube::golang::build_binaries_for_platform() {
     testpkg=$(dirname "${test}")
 
     mkdir -p "$(dirname "${outfile}")"
+
+    # TODO(saschagrunert): try to remove once go1.20.1 got released, watch
+    # out for the release notes which may fix relocation issues like:
```
[the remark this answers is not in the document]
Comment (liggitt) [reply hack/lib/golang.sh:785]: agreed, we need to report this ASAP to get it resolved in the earliest possible 1.20.x patch release

[the remark this answers is not in the document]
Comment (saschagrunert) [reply hack/lib/golang.sh:785]: Referencing the upstream issue https://github.com/golang/go/issues/58425

File: hack/lib/golang.sh (hunk 2 above)
Comment (saschagrunert) [reply hack/lib/golang.sh]: Changed as suggested by using a new local variable, which only applies to the `go test` statement.

**Comment by saschagrunert** [comment]
The `build-master` test is green again, because commit b0171f77f68427813efbb478c0ba89f48322d34a magically fixed the relocation problem. See https://github.com/kubernetes/kubernetes/issues/115605#issuecomment-1423874758, https://k8s-testgrid.appspot.com/sig-release-master-blocking#build-master

I think we don't need this PR immediately now, but can reopen if it occurs again.

File: hack/lib/golang.sh [hunk 4]
```diff
@@ -775,12 +775,28 @@ kube::golang::build_binaries_for_platform() {
...
+    # go.go:(.text+0x207ca94): relocation truncated to fit: R_ARM_CALL against `runtime.duffcopy'
+    #
+    # Upstream issue: https://github.com/golang/go/issues/58425
+    # Ref: https://github.com/kubernetes/kubernetes/issues/115605
+    if [[ "${test}" == k8s.io/kubernetes/test/e2e_node/e2e_node.test && "${platform}" == linux/arm ]]; then
+      kube::log::info "    disabling CGO for binary: ${test}"
+      use_cgo=0
+    fi
+
+    CGO_ENABLED="$use_cgo" go test -c \
```
Comment (liggitt) [inline hack/lib/golang.sh:799]: do we have the full expanded `go test` invocation (including all the goflags / gogcflags / goasmflags / goldflags / tags) that was breaking? that would help in giving the go team a reproducer

**Comment by saschagrunert** [comment]
Reopening because of https://github.com/kubernetes/kubernetes/issues/115675

**Comment by saschagrunert** [comment]
PTAL again @kubernetes/sig-release-leads @kubernetes/release-engineering 
hack/OWNERS: @BenTheElder @cblecker @dchen1107 @deads2k @spiffxp @thockin

File: hack/lib/golang.sh [hunk 5]
```diff
@@ -775,12 +775,31 @@ kube::golang::build_binaries_for_platform() {
   fi
 
   for test in "${tests[@]:+${tests[@]}}"; do
-    local outfile testpkg
+    local outfile testpkg use_cgo
     outfile=$(kube::golang::outfile_for_binary "${test}" "${platform}")
     testpkg=$(dirname "${test}")
+    use_cgo="${CGO_ENABLED:-0}"
```
Comment (liggitt) [inline hack/lib/golang.sh:781]: should the default be 0 or 1?

Comment (liggitt) [reply hack/lib/golang.sh:781]: the fact that we only override to 0 for things in the statics list above makes me think the default should be 1 here, right?

Comment (saschagrunert) [reply hack/lib/golang.sh:781]: The tests failed when the default was `1`: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/115613/pull-kubernetes-cross/1623945772418469888/build-log.txt

So I changed the default to `0`, which seems to be right in the context of the test binaries.

Comment (liggitt) [reply hack/lib/golang.sh:781]: if we're defaulting to 0, then setting to 0 below specifically for linux/arm doesn't seem like it's doing anything, right?

Comment (liggitt) [reply hack/lib/golang.sh:781]: we have kubelet and e2e_node tests that use cgo, so turning this off is short-circuiting those tests, right?

Comment (aojea) [reply hack/lib/golang.sh:781]: `-race`requires CGO to 1

Comment (saschagrunert) [reply hack/lib/golang.sh:781]: Ugh, yeah means we loose some tests when disabling CGO. :/

Comment (aojea) [reply hack/lib/golang.sh:781]: it simple does not work, 
```
CGO_ENABLED=0 go test -race .
go: -race requires cgo; enable cgo by setting CGO_ENABLED=1
```
I don't know if those tests set that variable later, it was more just a FYI

Comment (liggitt) [reply hack/lib/golang.sh:781]: looks like race isn't supported on linux/arm anyway... I think defaulting CGO_ENABLED to "" would be more reasonable, though we'd still lose those two cgo tests on linux/arm (though I'm not even sure why they need cgo, honestly)

Comment (liggitt) [reply hack/lib/golang.sh:781]: bisected this to the unified build getting enabled in go1.20, details in https://github.com/golang/go/issues/58425#issuecomment-1426346686

alternative workaround in https://github.com/kubernetes/kubernetes/pull/115692

**Comment by liggitt** [comment]
/close
In favor of https://github.com/kubernetes/kubernetes/pull/115692

**Pull request closed without merging**